promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent.
+	isNewLineApplied bool
+	// containsGrouping represents whether a binary expression contains
+	// a grouping modifier.
+	containsGrouping bool
+	// immediateScalar represents whether the immediate child of a binary
+	// expression is a scalar.
+	immediateScalar       bool
+	insideMultilineBraces bool
+	previousBaseIndent    string
+}
+
+// New returns a new prettier over the given slice of files.
+func New(Type uint, content interface{}) (*Prettier, error) {
+	var (
+		ok         bool
+		files      []string
+		expression string
+		typ        uint
+	)
+	switch Type {
+	case PrettifyRules:
+		files, ok = content.([]string)
+		typ = PrettifyRules
+	case PrettifyExpression:
+		expression, ok = content.(string)
+		typ = PrettifyExpression
+	}
+	if !ok {
+		return nil, errors.Errorf("invalid type: %T", reflect.TypeOf(content))
+	}
+	return &Prettier{
+		files:      files,
+		expression: expression,
+		Type:       typ,
+		pd:         newPadder(),
+	}, nil
+}
+
+func newPadder() *padder {
+	return &padder{
+		indent:             "  ", // 2 space
+		isNewLineApplied:   true,
+		columnLimit:        100,
+		previousBaseIndent: "  ", // start with initial indent as previous val to skip new line at start.
+	}
+}
+
+// Run executes the prettier over the rules files or expression.
+func (p *Prettier) Run() []error {
+	var (
+		groupFiles []*rulefmt.RuleGroups
+		errs       []error
+	)
+	switch p.Type {
+	case PrettifyRules:
+		for _, f := range p.files {
+			ruleGroups, err := p.parseFile(f)
+			if err != nil {
+				for _, e := range err {
+					errs = append(errs, errors.Wrapf(e, "file: %s", f))
+				}
+			}
+			groupFiles = append(groupFiles, ruleGroups)
+		}
+		if errs != nil {
+			return errs
+		}
+		for _, rgs := range groupFiles {
+			for _, grps := range rgs.Groups {
+				for _, rules := range grps.Rules {
+					exprStr := rules.Expr.Value
+					standardizeExprStr := p.expressionFromItems(p.lexItems(exprStr))
+					if err := p.parseExpr(standardizeExprStr); err != nil {
+						return []error{err}
+					}
+					formattedExpr, err := p.prettify(
+						p.sortItems(p.lexItems(exprStr)),
+						0,
+						"",
+					)
+					if err != nil {
+						return []error{errors.Wrap(err, "Run")}
+					}
+					fmt.Println("output is ")
+					fmt.Println(formattedExpr)
+				}
+			}
+		}
+
+	}
+	return nil
+}
+
+func (p *Prettier) sortItems(items []parser.Item) []parser.Item {
+	for i := 0; i < len(items); i++ {
+		item := items[i]
+		switch item.Typ {
+		case parser.SUM, parser.AVG, parser.MIN, parser.MAX, parser.COUNT, parser.COUNT_VALUES, parser.STDDEV,
+			parser.STDVAR, parser.TOPK, parser.BOTTOMK, parser.QUANTILE:
+			var (
+				formatItems       bool
+				aggregatorIndex   = i
+				keywordIndex      = -1
```

Comment:
These magic numbers reduce code clarity. Booleans and numbers might be clearer.

Reply:
You mean the entire block, or `*Index` only?

Reply:
The latter.

Reply:
We need to store the index of the keywords and the parens, so anyways `int` is a requirement.

However, I didn't get _Booleans and numbers might be clearer_. Can you please explain it?

Reply:
Instead of encoding the special case of `keywordIndex == -1` using the magic value `-1`, use another variable for encoding that state.

Reply:
Addressed in recent commits. Hence, resolving.